Fix: Retain playback panel positioning on restarting application. Vis…

worldwideweary opened the pull request:
Resolves: https://github.com/Jojo-Schmitz/MuseScore/issues/1040

Attempts to (edited):
- Retain geometry of _Play Panel_ and keeps its float/docked state
- _Play Panel_ honors the Preference in General to show or not on startup

Testing phase

worldwideweary commented on mscore/musescore.cpp:
```diff
@@ -8349,10 +8349,13 @@ void MuseScore::init(QStringList& argv)
             qApp->processEvents();
             }
 
-      mscore->showPlayPanel(preferences.getBool(PREF_UI_APP_STARTUP_SHOWPLAYPANEL));
```
Hey. Didn't even notice the option in the preferences (General tab in English). This change overrides/negates it. so will have to update this in some way. This makes it only show if it was open the last time the app was open, at least that is what it should do.

I'm not sure the best way about it... they seem kind of irreconcilable though: either have it be hard coded with that preference as in 3.6.2, or be dynamic as with like the inspector/palette. I can always revert this portion of the code back so that it isn't dynamic but follows the old way if that seems more agreeable. Either that or remove that option completely. I kind of like it being dynamic without a preference, but don't have any personal feelings about it either way because I'm used to activating it with a shortcut if need be.

worldwideweary replied:
Which would you prefer: have the play panel show on start-up based on if it was open last (new behavior), or just keep it based on the option in the preferences (old)?

worldwideweary replied:
Yet what would that mean if the user goes to the preferences and unsets it explicity, but then keeps the play panel open and closes the app and restarts? The  unsetting would have been overriden by the fact that it was open at last session shutdown...

Jojo-Schmitz replied:
Hmm, maybe with having that preference we don't need that last session stuff at all.
It should just remember whether docked or undocked

worldwideweary replied:
Yeah, that would be the default thing to do. I can change the PR to just follow the brute force always show at start or always not show as it is. I think I'll keep the dynamic stuff like with the other panels for my own build though if we go with that.

Jojo-Schmitz replied:
We could OR the preference and the last open

worldwideweary replied:
True, but it feels kind of funny to have "Play Panel" off on startup for the setting, yet still have it show up because the user didn't close it manually. It's an "imperfect logical situation" in my opinion trying to combine them.

worldwideweary replied:
If we used OR, wouldn't it be like:
1) Preference On: Show always
2) Preference Off: Show only when last time shown

Jojo-Schmitz replied:
No. Preference decides whether to show always.
Open last time decides whether to open next time, even if off in preferences.

Jojo-Schmitz replied:
Gives you the flexibility you want without rendering the preference useless

Jojo-Schmitz replied:
yep

worldwideweary commented:
Updated: removed saving visibility but honor the startup preference to always be shown or not shown.
Should just retain positioning which now includes positioning + floating status hopefully.

worldwideweary commented:
Problem: apparently it's not retaining its current dock state. It seems to always be floating on load. Will look into it later

worldwideweary commented:
Turns out for some reason that the float-state is always set to true (hence was always manually set to false in the code it seems) on start from what I can tell, so I saved the float state and also actively restored geometry so that not only is the "size" of the floating state saved, but also its positioning (don't think that was saved earlier). Should be good to go, but further testing etc...

Jojo-Schmitz commented on mscore/musescore.cpp:
```diff
@@ -8349,10 +8354,15 @@ void MuseScore::init(QStringList& argv)
+
+      settings.beginGroup("MainWindow");
+      const bool forceShowPlayPanel = preferences.getBool(PREF_UI_APP_STARTUP_SHOWPLAYPANEL);
+      const bool lastSessionShowedPlayPanel = settings.value("showPlayPanel").toBool();
```
Don't those `const' mean that these values won't ever change in the current session after the `init()`?
